### PR TITLE
Geocoding/refactoring

### PR DESCRIPTION
**Pull request by Sascha-Gschwind**
### CATEGORY

Choose one

- [ ] Bug Fix
- [ ] Enhancement (new features, refinement)
- [x] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/utils/geocoders.py [hunk 1]
```diff
@@ -58,15 +67,23 @@ def _append_cords_to_data_entry(self, data_entry: list, geocoded: list):
...
         self.progress["success_counter"] = 0
         self.progress["doubt_counter"] = 0
         self.progress["failed_counter"] = 0
         self.progress["is_in_progress"] = in_progress
         self.progress["progress"] = 0
 
-    def _set_result_precision_counters(self):
+    def _set_result_precision_counters(self) -> dict:
+        """
+        Sets the success, doubt and faled counters
```
Comment (sloypog) [inline superset/utils/geocoders.py:86]: failed

File: superset/utils/geocoders.py [hunk 2]
```diff
@@ -58,15 +67,23 @@ def _append_cords_to_data_entry(self, data_entry: list, geocoded: list):
         data_entry.append(str(coordinates[1]))
         return data_entry
 
-    def _set_initial_states(self, in_progress=False):
+    def _set_initial_states(self, in_progress=False) -> None:
+        """
+        Sets the initial state of the progress
```
Comment (sloypog) [inline superset/utils/geocoders.py:74]: progress dictionary?

File: superset/utils/geocoders.py [hunk 3]
```diff
@@ -136,9 +157,18 @@ def _geocode(self, data: list) -> list:
         return ["", [geocoded_data, self._set_result_precision_counters()]]
 
     def _get_coordinates_from_address(self, address: str):
```
Comment (sloypog) [inline superset/utils/geocoders.py:163]: add return value as optional[list] or does that fail?

File: superset/views/geocoding.py [hunk 4]
```diff
@@ -134,76 +137,66 @@ def geocode(self) -> Response:
...
+                lat_column, lat_exists, lon_column, lon_exists, table
             )
+            message_suffix = f" but columns have been created please choose the 'Replace' or 'Append' option when trying again."
+            table_data = self._load_data_from_columns(table, columns)
+            table_data = self._filter_data(table_data, len(columns) - 1, len(columns), "append" in if_exists)
+        except (NoColumnsException, TableNotFoundException, SqlSelectException) as e:
             self.stats_logger.incr("geocoding_failed")
-            return json_error_response(f"{e.args[0]} {message_suffix}", status=500)
+            self.logger.exception(f"Error while verifying parameters {e}")
+            return json_error_response(e.args[0], status=400)
```
Comment (sloypog) [inline superset/views/geocoding.py:165]: why is message_suffix not returned here?

File: superset/views/geocoding.py [hunk 5]
```diff
@@ -134,76 +137,66 @@ def geocode(self) -> Response:
...
+            table = self._get_table_with_columns(table_id)
+            lat_exists = self._does_column_name_exist(table, lat_column)
+            lon_exists = self._does_column_name_exist(table, lon_column)
+            if "fail" in if_exists and (lat_exists or lon_exists):
+                self.stats_logger.incr("geocoding_failed")
+                return json_error_response("At least one of the columns already exists!", status=400)
+            self._create_columns(
+                lat_column, lat_exists, lon_column, lon_exists, table
             )
+            message_suffix = f" but columns have been created please choose the 'Replace' or 'Append' option when trying again."
```
Comment (sloypog) [inline superset/views/geocoding.py:158]: what if both columns already exist? And also if replace/append was chosen we still get this message?

File: superset/views/geocoding.py [hunk 6]
```diff
@@ -212,7 +205,7 @@ def geocode(self) -> Response:
                 f"Failed to add columns/ data after geocoding {e.orig}"
```
Comment (sloypog) [inline superset/views/geocoding.py:203]: shouldn't this be "failed to add data..." only as the columns are created before geocoding? (Specifically mentioned in the documentation)

File: superset/views/geocoding.py [hunk 7]
```diff
@@ -230,200 +223,174 @@ def geocode(self) -> Response:
...
+        if street_column:
+            columns.append(street_column)
+        if city_column:
+            columns.append(city_column)
+        if country_column:
+            columns.append(country_column)
+        if building_number_column:
+            columns.append(building_number_column)
+        columns.append(request_data.get("latitudeColumnName", "lat"))
+        columns.append(request_data.get("longitudeColumnName", "lon"))
```
Comment (sloypog) [inline superset/views/geocoding.py]: this is already done in the method calling this one, but I'm guessing that's also part of the refactoring still going on?

File: superset/views/geocoding.py [hunk 8]
```diff
@@ -230,200 +223,174 @@ def geocode(self) -> Response:
...
-                    "Column name {0} for latitude is already in use".format(lat_column)
-                )
-            if lon_exists:
-                raise ValueError(
-                    "Column name {0} for longitude is already in use".format(lon_column)
-                )
-            self._add_lat_lon_columns(table, lat_column, lon_column)
-            column_message_suffix = (
-                f"successfully created columns: {lat_column} and {lon_column}"
+    def _create_columns(
```
Comment (sloypog) [inline superset/views/geocoding.py:258]: docstring still says check and create but name implies it only creates without any checking?

File: superset/views/geocoding.py [hunk 9]
```diff
@@ -230,200 +223,174 @@ def geocode(self) -> Response:
...
+        self, lat_column: str, lat_exists: bool, lon_column: str, lon_exists: bool, table: SqlaTable
+    ) -> None:
+        """
+        Check and create the lat / lon columns if needed
+        :param if_exists: What should happen if the columns already exist
+        :param lat_column: The name of the latitude column
+        :param lon_column: The name of the longitude column
+        :param table_id: The id of the table
+        :return: A message describing what columns have been created or an empty message if no columns have been created
```
Comment (sloypog) [inline superset/views/geocoding.py]: there is no param if_exists (not needed, remove from docstring), params lat_exists and lon_exists are not mentioned in docstring
docstring mentions table_id but signature contains table

File: superset/views/geocoding.py [hunk 10]
```diff
@@ -230,200 +223,174 @@ def geocode(self) -> Response:
...
 
     def _load_data_from_columns(
-        self, id: int, columns: list, lat_column=None, lon_column=None
-    ):
+        self, table: SqlaTable, columns: list
+    ) -> list:
         """
         Get data from columns form table
-        :param table_name: The table name from table from which select
+        :param table_id: The id of the table
```
Comment (sloypog) [inline superset/views/geocoding.py]: signature changes from id to table object but docstring changes from table-object to table id?

File: superset/views/geocoding.py [hunk 11]
```diff
@@ -230,200 +223,174 @@ def geocode(self) -> Response:
...
-        self, table: SqlaTable, lat_column: str = None, lon_column: str = None
-    ):
-        """
-        Add new longitude and latitude columns to table
-        :param table: The table insert columns
-        :param lat_column: The name of latitude column
-        :param lon_column: The name of longitude column
-        :raise SqlAddColumnException: When add column-SQL went wrong
+        :param data: The data to be geocoded as a list of tuples
+        :return: A list of tuples containing the data and the corresponding long, lat values
```
Comment (sloypog) [inline superset/views/geocoding.py:366]: lat,lon

File: superset/views/geocoding.py (hunk 5 above)
Comment (Sascha-Gschwind) [reply superset/views/geocoding.py:158]: A lot of code was written only for this message, all the user needs to know is that the "Fail" option is no good anymore

**Pull request closed without merging**